File: api/src/endpoint/repository/create_repository.rs
use serde::{Deserialize, Serialize};

use super::default_visibility;
use crate::{endpoint::Endpoint, resource::RepositoryResource};

pub struct CreateRepository;

impl Endpoint for CreateRepository {
    const PATH: &'static str = "/repository/{owner}/{repo}";
    const METHOD: http::Method = http::Method::POST;

    type Request = CreateRepositoryRequest;
    type Response = CreateRepositoryResponse;
}

#[derive(ApiRequest, Debug, Serialize, Deserialize)]
pub struct CreateRepositoryRequest {
    pub owner_type: String,

    #[serde(default = "default_visibility")]
    pub visibility: String,
}

pub type CreateRepositoryResponse = RepositoryResource;

File: gitdot-api-ts/src/endpoint/repository/create-repository.ts
import { z } from "zod";

import { RepositoryResource } from "../../resource";

export const CreateRepositoryRequest = z.object({
  owner_type: z.string(),
  visibility: z.string(),
});
export type CreateRepositoryRequest = z.infer<typeof CreateRepositoryRequest>;

export const CreateRepository = {
  path: "/repository/{owner}/{repo}",
  method: "POST",
  request: CreateRepositoryRequest,
  response: RepositoryResource,
} as const;
export type CreateRepository = typeof CreateRepository;
